src/mbgl/map/annotation.cpp

```diff
@@ -77,118 +67,314 @@ uint32_t AnnotationManager::nextID() {
+                        std::vector<std::vector<vec2<double>>>({{ pp }}),
+                        AnnotationType::Point,
+                        {{ }},
+                        pointFeatureProperties,
+                        maxZoom
+                    );
+
+                    std::copy(featureAffectedTiles.begin(), featureAffectedTiles.end(), std::inserter(affectedTiles, affectedTiles.begin()));
+
+                    annotationIDs.push_back(pointAnnotationID);
+                }
+            }
+        } else {
+
+            const uint32_t shapeAnnotationID = nextID();
+
+            // current shape tiles are on-the-fly, so we don't get any "affected tiles"
+            // and just expire all annotation tiles for shape adds
+
+            addTileFeature(
+                shapeAnnotationID,
+                segments[s],
+                {{ }},
+                AnnotationType::Shape,
+                styleProperties[s],
+                {{ }},
+                maxZoom
+            );
+
+            annotationIDs.push_back(shapeAnnotationID);
+        }
+    }
+
+    // Tile:IDs that need refreshed and the annotation identifiers held onto by the client.
+    return std::make_pair(affectedTiles, annotationIDs);
+}
 
-    std::vector<uint32_t> annotationIDs;
-    annotationIDs.reserve(points.size());
+std::unordered_set<TileID, TileID::Hash>
+AnnotationManager::addTileFeature(const uint32_t annotationID,
+                                  const AnnotationSegments& segments,
+                                  const std::vector<std::vector<vec2<double>>>& projectedFeature,
+                                  const AnnotationType& type,
+                                  const StyleProperties& styleProperties,
+                                  const std::unordered_map<std::string, std::string>& featureProperties,
+                                  const uint8_t maxZoom) {
 
-    std::vector<TileID> affectedTiles;
+    assert(type != AnnotationType::Any);
 
-    for (size_t i = 0; i < points.size(); ++i) {
-        const uint32_t annotationID = nextID();
+    // track the annotation global ID and its original geometry
+    auto anno_it = annotations.emplace(annotationID,
+        std::make_unique<Annotation>(type, segments, styleProperties));
 
-        // track the annotation global ID and its geometry
-        auto anno_it = annotations.emplace(
-            annotationID,
-            std::make_unique<Annotation>(AnnotationType::Point,
-                                          AnnotationSegments({ { points[i] } })));
+    std::unordered_set<TileID, TileID::Hash> affectedTiles;
 
-        const uint8_t maxZoom = data.transform.getMaxZoom();
+    if (type == AnnotationType::Shape) {
 
-        // side length of map at this zoom
+        orderedShapeAnnotations.push_back(annotationID);
+
+        using namespace mapbox::util::geojsonvt;
+
+        const uint32_t z2 = 1 << 14;
+        const double baseTolerance = 3;
```

Comment:
it's a good default for simplification in my tests, although it should be configurable (it is in JS)